### PR TITLE
Maj

### PR DESCRIPTION
**Pull request by Ydonse**

### CODE REVIEW

File: commands/commands.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module.exports = class command {
```
Comment (johanlajili) [inline commands/commands.js]: class names should start with an uppercase

File: stage/messages.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+const prefix = process.env.PREFIX;
+const cmdName = "stage";
+
+module.exports = class MSG {
```
Comment (johanlajili) [inline stage/messages.js:25]: image

File: commands/stage.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+let currentUsers = [];
+
+module.exports = class Stage extends Command {
+  static match(message) {
+    if (!this.checkCommand(message) || message.author.bot) return false;
+    return true;
+  }
+
+  static async action(message) {
+    if (!User.checkUsersnb(message)) return Msg.usage(message);
```
Comment (johanlajili) [inline commands/stage.js]: needed to dig around to undersand the logic here.  Msg.usage does not return anything also so it's a bit confusing to return it.

I'd suggest

```javascript
if (!Stage.validateCommand(message)){
   return
}
[...]
static validateCommand(message){
     const numberOfUserIsCorrect = User.checkUsersNumber(message);
     return numberOfUserIsCorrect; // can be extended to more checks in the fure.
}

Comment (johanlajili) [reply commands/stage.js]: I would also move that logic
```
let user1 = User.fromMention(args[0]);
    let user2 = User.fromMention(args[1]);
    if (
      User.isUserValid(message, user1, currentUsers) !== true ||
      User.isUserValid(message, user2, currentUsers) !== true
    )
      return false;
    else if (user1 != message.author && user2 != message.author)
      return Msg.usage(message);
      ```
and similar
      to be inside that initial test.

File: commands/stage.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    currentUsers.push(user1, user2);
+    await Selector.chooseStage(
+      user1,
+      user2,
+      this.getRound(args),
+      message
+    ).catch((error) => console.log(error));
+    currentUsers = currentUsers.filter(function (value) {
+      return value != user1 && value != user2;
+    });
```
Comment (johanlajili) [inline commands/stage.js]: huh?

File: commands/stage.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    return message.content.startsWith(prefix + cmdName);
+  }
+
+  static getRound(args) {
+    //returns round of the match
+    if (
+      args.length > 2 &&
+      ["2", "3", "tour2", "tour3"].includes(args[2].toLowerCase())
+    )
+      return 2;
```
Comment (johanlajili) [inline commands/stage.js:53]: bit confusing that tour3 / 3 are returning 2.

File: main.js [hunk 6]
```diff
@@ -1,209 +1,21 @@
...
-          currentUsers.push(args[0], args[1]);
-          await chooseStage(user1, user2, args, message);
-          currentUsers = currentUsers.filter(function (value, index, arr){
-            return value != args[0] && value != args[1];
-          });
-        }
-      }
-    }
+  const img = IMG.getImages();
+  img ? Stage.parse(message) : !message.author.bot && MSG.imagesLoading(message);
```
Comment (johanlajili) [inline main.js]: It's generally poor practice to use ternaries for instructions. Ternaries return something so they are generally use in context where the return value is used.
Here use a simple if / else.

File: stage/canvas.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+          ctx.drawImage(
+            images[8],
+            espacementX * 3 + sizex * 2,
+            espacementY * 3 + sizey * 2,
+            sizex,
+            sizey
+          );
+        }
+      }
+
```
Comment (johanlajili) [inline stage/canvas.js]: Une petite for loop ferait un peu plus professionel ici

```javascript
for (let i; i< 3; i++){
    for (let j; j < 3; j++){
        const imageIndex = i*j;
            if (levels.includes((imageIndex + 1).toString())) {
                ctx.drawImage(
                  images[imageIndex],
                  espacementX * (i+1) + sizex * i,
                  espacementY * (j+1) + sizey * j,
                  sizex,
                  sizey
                );
              }
     }
}
```

File: stage/images.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  0: "http://www.donsefactory.com/wp-content/uploads/2021/02/DF-2.jpg",
+  1: "http://www.donsefactory.com/wp-content/uploads/2021/02/CB-2.jpg",
+  2: "http://www.donsefactory.com/wp-content/uploads/2021/02/SB-2.jpg",
+  3: "http://www.donsefactory.com/wp-content/uploads/2021/02/Smashville-2.jpg",
+  4: "http://www.donsefactory.com/wp-content/uploads/2021/02/Kalos-2.jpg",
+  5: "http://www.donsefactory.com/wp-content/uploads/2021/02/Ville-2.jpg",
+  6: "http://www.donsefactory.com/wp-content/uploads/2021/02/Yoshi-2.jpg",
+  7: "http://www.donsefactory.com/wp-content/uploads/2021/02/SP2-2.jpg",
+  8: "http://www.donsefactory.com/wp-content/uploads/2021/02/Lylat-2.jpg",
+};
```
Comment (johanlajili) [inline stage/images.js]: Normalement tu peux aussi mettre tes images directement dans le dossier de ton script, ça doit demander un peu de réflexion mais faisable et plus propre. Si demain tu change ton site par example ça te bousillerais ce projet.

File: stage/messages.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  static usage(message) {
+    message.reply(
+      `usage : \`${prefix}${cmdName} @participant1 @participant2 [option : tour2]\`\n For any questions ask Steyo`
+    );
+  }
+  static getMessageToSend(userWanted, levelsNbToBan) {
+    let fr = this.checkLanguage(userWanted);
```
Comment (johanlajili) [inline stage/messages.js]: Bel effort pour la localisation.
En général, une implémentation de localisation se fait avec dictionnaire de clé pour être plus lisible.

```javascript
const localisation = {
   fr: {
        "ban" : "{userWanted} bannis le dernier stage"
         "send": "{userWanted} Envoie le/les numéro du stage que tu veux banir"
    },
    en: {
         "ban": "{userWanted} ban the last stage, {remaining} restant",
         "send": "{userWanted} send the stage number you want to ban {remaining} left"
    }
}

getMessageToSend(userWanted, levelsNbToBan) {
    if (levelsNbToBan[0] == 1 && levelsNbToBan.length == 1) {
        return localisation[language].ban.replace('{userWanted}', userWanted);
    }
    return localsiatiton[language].send.replace('{userWanted}, userWanted).replace('{remaining}', levelsNbToBan[0])
    
}

File: stage/selector.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    let userWanted;
+    let reponses;
+    let levelsNbToBan = round === 1 ? [3, 4, 1] : [3, 5];
+    const canvas = Canvas.createCanvas(1000, 700);
+
+    while (levelsRemaining > 1) {
+      let attachment;
+      userWanted = User.getUserWanted(user1, user2, round, levelsRemaining);
+      if (levelsRemaining == 9) attachment = { files: [allStages] };
+      else {
```
Comment (johanlajili) [inline stage/selector.js]: just stylistic but if you're gonna have an else, you better use curly brackets on your if.

File: stage/user.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    const id = matches[1];
+    return client.users.cache.get(id);
+  }
+
+  static getUserWanted(user1, user2, tour, levelsRemaining) {
+    //return user whose answer is awaited
+    if (tour == 1) {
+      if (levelsRemaining > 6) return user1;
+      else if (levelsRemaining > 2) return user2;
+      else return user1;
```
Comment (johanlajili) [inline stage/user.js]: Personally I'd extract all this logic about turns / levelsRemaining into a separate file, which would in essence contain your core logic.

**Review by johanlajili — COMMENTED** [review]
Pas trop mal. En général je te conseille d'ajouter prettier et eslint à ton projet, ça fait pas de mal.

J'ai mis pas mal de commentaire, libre à toi de les suivres ou non 🤷‍♂️

File: commands/stage.js (hunk 5 above)
Comment (Ydonse) [reply commands/stage.js:53]: It's juste that the same rule applies, but it's invisible for the user.

File: commands/stage.js (hunk 4 above)
Comment (Ydonse) [reply commands/stage.js]: It removes user1 and 2 from currentUsers. What is confusing you ?

Comment (johanlajili) [reply commands/stage.js]: but isn't current user only containing user1 and 2? If not then there's probably more thinking to do surrounding how multiple users interact with one another

**Pull request merged**